core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockPusher.scala

```diff
@@ -251,6 +251,10 @@ private[spark] class ShuffleBlockPusher(conf: SparkConf) extends Logging {
         }
         handleResult(PushResult(blockId, exception))
       }
+
+      override def onSaslTimeout(): Unit = {
+        TaskContext.get().taskMetrics().incSaslRequestRetries(1)
```

Comment:
This is not within the task thread - you should be getting an NPE as task context should be null.

If the tests are not failing, we have to make sure we are testing this fix appropriately - some codepaths (like push related suite's) might not be getting tested with this ?

Reply:
In our internal fork, ShuffleBlockPusher is created with the `writeMetrics`. We can change the constructor to use `taskMetrics`. However, then the concern is that push is happening outside the task thread so whether the metric is correct or not. One way to move forward is to make this metric count SASL retries on when shuffle data is fetched and not pushed.

Reply:
Updating metrics once the task has terminated does not help - since they are not propagated to driver.
One option would be to rely on `ExecutorMetricsSource` ?

Reply:
discussed offline, we will not making the metric change as no other network metric is exposed anyways. If we need it down the line, we can add it as a follow-up task.